presto-hive/src/test/java/com/facebook/presto/hive/TestHiveMaterializedViewLogicalPlanner.java

```diff
@@ -1471,6 +1474,296 @@ public void testMaterializedViewForJoin()
+
+            assertUpdate(format("CREATE MATERIALIZED VIEW %s " +
+                            "WITH (partitioned_by = ARRAY['ds']) " +
+                            "AS SELECT suppkey, name, ds\n " +
+                            "FROM %s " +
+                            "WHERE name != 'bob'",
+                    suppliersView, supplierTable));
+
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-11'", lineItemView1), 100);
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-12'", lineItemView1), 100);
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-11'", lineItemView2), 100);
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-12'", lineItemView2), 100);
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-11'", suppliersView), 100);
+            assertUpdate(format("REFRESH MATERIALIZED VIEW %s WHERE ds='2021-07-12'", suppliersView), 100);
+
+            assertTrue(getQueryRunner().tableExists(getSession(), lineItemView1));
+            assertTrue(getQueryRunner().tableExists(getSession(), lineItemView2));
+            assertTrue(getQueryRunner().tableExists(getSession(), suppliersView));
+
+            setReferencedMaterializedViews((DistributedQueryRunner) queryRunner, lineItemTable, ImmutableList.of(lineItemView1, lineItemView2));
+            setReferencedMaterializedViews((DistributedQueryRunner) queryRunner, supplierTable, ImmutableList.of(suppliersView));
+
+            String baseQuery = format("WITH long_name_supp AS ( \n" +
+                    "SELECT suppkey, name \n" +
+                    "FROM %s \n" +
+                    "WHERE name != 'bob'), " +
+                    "supp_max AS (\n" +
+                    "SELECT suppkey, MAX(quantity) AS max_qty \n" +
+                    "FROM %s \n" +
+                    "GROUP BY suppkey, ds), \n" +
+                    "supp_sum AS (\n" +
+                    "SELECT suppkey, SUM(quantity) AS qty \n" +
+                    "FROM %s \n" +
+                    "GROUP BY suppkey, ds) \n" +
+                    "SELECT n.suppkey, n.name, m.max_qty, s.qty\n " +
+                    "FROM long_name_supp AS n \n" +
+                    "LEFT JOIN supp_max AS m ON n.suppkey = m.suppkey \n" +
+                    "LEFT JOIN supp_sum AS s ON n.suppkey = s.suppkey \n" +
+                    "ORDER BY suppkey, name",
+                    supplierTable, lineItemTable, lineItemTable);
+
+            MaterializedResult optimizedQueryResult = computeActual(queryOptimizationWithMaterializedView, baseQuery);
+            MaterializedResult baseQueryResult = computeActual(baseQuery);
+            assertEquals(baseQueryResult, optimizedQueryResult);
+
+            PlanMatchPattern expectedPattern = anyTree(
+                    node(JoinNode.class,
+                            node(JoinNode.class,
+                                    exchange(
```

Comment:
-It seems the exchange and join nodes are necessary to match leaves with anytree if there is an intermediate node with != 1 sources- if this isn't fully correct I'd like to know.
-I tried putting multiple non-leaf `PlanMatchPattern`s on the same line (e.g. `exchange(anyTree(exchange(` but this ultimately felt less readable. I know we generally prefer a more compact code style though, and I don't feel strongly about it either way if someone else does.